tests/test_geoserver.py

```diff
@@ -46,17 +60,89 @@ def get_geoserver_settings():
     return geoserver_settings
 
 
+def prepare_geoserver_test_workspace(test_instance, geoserver_handler, workspace_key):
+    """
+    Prepares a workspace, its datastore and a test shapefile along with the associated Geoserver resources.
+    """
+    # Creating workspace and datastore needed to publish a shapefile
+    workspace_name = test_instance.workspaces[workspace_key]
+    datastore_name = f"test-datastore-{workspace_key}"
+    geoserver_workspace_path = f"/user_workspaces/{workspace_name}/shapefile_datastore"
+    geoserver_handler._create_workspace_request(workspace_name=workspace_name)
+    geoserver_handler._create_datastore_request(workspace_name=workspace_name,
+                                                datastore_name=datastore_name)
+    geoserver_handler._configure_datastore_request(workspace_name=workspace_name,
+                                                   datastore_name=datastore_name,
+                                                   datastore_path=geoserver_workspace_path)
+
+    # Preparations needed to make tests work without all the other handlers running
+    datastore_path = get_datastore_path(test_instance.workspace_folders[workspace_key])
+
+    # This next part can fail if the user running this test doesn't have write access to the directory
+    copy_shapefile(basename=test_instance.test_shapefile_name, destination=datastore_path)
```

Comment:
Is it possible to try/except some error, or do a precheck, to reraise a relevant error message directly? Or is the failure sufficiently descriptive already?

Should `reset_geoserver_test_workspace` be called beforehand to ensure a consistent initial condition?
If tests are abruptly stopped, could rerunning them generate some kind of edge-case problem because previous cleanup did not complete?

Reply:
I think the failure will be descriptive enough. It should trigger a `PermissionError` / `Permission denied: <path>` exception.

The teardown for the main class already calls the `reset_geoserver_test_workspace` function to clear things up if the tests fail. I think if the user does something like running a test while another test is paused (while debugging), it might cause an error, but stopping the test should clear the workspace...